README.md
# enjoy-version

A version compare tool, which is useful in comparing versions of APPs and JavaScript modules.

一个版本比较工具，在APP、JavaScript模块的等的版本比较中非常有用。

## 安装

```bash
npm i enjoy-version
```

## 引入

```javascript
import { Version, libVersion } from 'enjoy-version';
```
## API

### libVersion

获得当前类库的版本号。

### Version

是一个类，可以用来创建一个版本对象，例如`new Version('1.2.3')`，将获得一个对象：`{"version":"1.2.3"}`。

`new Version('1.2.3')`产生的对象还有`gt(version)`、`gte(version)`、`lt(version)`、`lte(version)`、`eq(version)`这几个方法，可以将`new Version('1.2.3')`中的version属性 与 传入的version（版本字符串，如'1.0.0'）参数进行比较，判断版本的大小。返回的值是Boolean类型。

- `gt(version)`表示大于。

- `gte(version)`表示大于等于。

- `lt(version)`表示小于。

- `lte(version)`表示小于等于。

- `eq(version)`表示等于。

例如：

`new Version('1.2.3').gt('1.0.0')`为true;

`new Version('1.2.3').gte('1.2.0')`为true;

`new Version('1.2.3').lt('1.2.4')`为true;

`new Version('1.2.3').lte('1.2.4')`为true;

`new Version('1.2.3').eq('1.2.3')`为true。


CHANGELOG.md
# enjoy-version 更新记录

## 1.0.0

* 首次发布


test/index_test.js
/**
 * assert(value[, message]) 相当于assert.ok(value[, message])
 * assert.strictEqual(actual, expected[, message])
 * assert.equal(actual, expected[, message])
 * assert.notEqual(actual, expected[, message])
 * assert.strictEqual(actual, expected[, message])
 * assert.notStrictEqual(actual, expected[, message])
 * assert.deepEqual(actual, expected[, message])
 * assert.notDeepEqual(actual, expected[, message]) // 判断两个对象是否不是deeply equal
 * assert.deepStrictEqual(actual, expected[, message])
 * assert.notDeepStrictEqual(actual, expected[, message])
 * assert.throws(block[, error][, message])
 * assert.doesNotThrow(block[, error][, message])
 * assert.fail(actual, expected, message, operator)
 * assert.ifError(value) // Throws value if value is truthy.
 */
import chai from 'chai'; // assert是Node.js自带的断言模块
import { Version, libVersion } from '../src/index.js';

import jsdom from 'mocha-jsdom';

chai.should();

describe('enjoy-version', () => {
    describe('libVersion', () => {
        it('返回当前库的版本号正确', () => {
            // enjoyVersion.libVersion.should.equal('1.0.0');
        });
    });

    describe('Version()', () => {
        it('传入版本号为0，返回string类型的0', () => {
            new Version('0').version.should.equal('0');
        });

        it('传入合法的版本号，返回指定的版本号', () => {
            new Version('1.2.3').version.should.equal('1.2.3');
        });

        it('传入非法的版本号，返回空字符串', () => {
            new Version(undefined).version.should.equal('');
            new Version(NaN).version.should.equal('');
            new Version(null).version.should.equal('');
            new Version({ a: 123 }).version.should.equal('');
        });

        it('高版本大于低版本', () => {
            new Version('1.2.3').gt('0.2.3').should.equal(true);
        });

        it('低版本小于高版本', () => {
            new Version('1.2.3').lt('2.2.3').should.equal(true);
        });

        it('相同的版本应该相等', () => {
            new Version('1.2.3').eq('1.2.3').should.equal(true);
        });

        it('高版本大于等于低版本', () => {
            new Version('1.2.3').gte('0.2.3').should.equal(true);
        });

        it('某版本大于等于其相同版本', () => {
            new Version('1.2.3').gte('1.2.3').should.equal(true);
        });

        it('低版本小于等于高版本', () => {
            new Version('1.2.3').lte('2.2.3').should.equal(true);
        });

        it('某版本小于等于其相同版本', () => {
            new Version('1.2.3').lte('2.2.3').should.equal(true);
        });
    });
});


test/karma.config.js
/**
 *
 * @author paian
 * @email  pai_an@qq.com
 * @since  17/3/31
 */

'use strict';

var babel = require('rollup-plugin-babel');
var babelrc = require('babelrc-rollup').default;
var globals = require('rollup-plugin-node-globals');
var builtins = require('rollup-plugin-node-builtins');
var nodeResolve = require('rollup-plugin-node-resolve');
var istanbul = require('rollup-plugin-istanbul');
var eslint = require('rollup-plugin-eslint');
var replace = require('rollup-plugin-replace');
// Rollup的模块引用只支持 ES6 Module，其他的需要采用 npm 和 commonjs 的插件去解决
var json = require('rollup-plugin-json');// allows Rollup to import data = requirea JSON file
var commonjs = require('rollup-plugin-commonjs');// the majority of packages on npm are exposed as CommonJS modules. We need to convert CommonJS to ES2015 before Rollup can process them.

var pkg = require('../package.json');
var external = Object.keys(pkg.dependencies);
var testFiles = [
    'test/**/*_test.js',
    {
        pattern: 'src/**/*.js',
        included: false // false 表示初始化的时候不会使用 script 标签直接将相关 js 引入到浏览器，需要自己写代码加载
    }
];

module.exports = function(config) {
    config.set({
        // 基础路径，将会用在解析所有patterns (eg. files, exclude)
        basePath: '',

        // 使用到的框架
        // 目前支持的框架: https://npmjs.org/browse/keyword/karma-adapter
        frameworks: ['mocha'],

        // 将会在浏览器中执行的代码 list of files / patterns to load in the browser
        files: testFiles,

        // 需要从files中排除掉的文件
        exclude: [],

        // 在提供给浏览器之前需要做预处理的文件，以及这些文件对应的预处理器。
        // 对于ES6、typescript、coffee 等的编译都在这里进行
        // 目前支持的预处理器: https://npmjs.org/browse/keyword/karma-preprocessor
        preprocessors: {
            'test/**/*_test.js': ['rollup']
        },

        // rollup与处理器的配置
        rollupPreprocessor: {
            plugins: [
                json(),
                // https://www.npmjs.com/package/rollup-plugin-eslint
                eslint({
                    throwError: true,
                    exclude: ['package.json', 'node_modules/**']
                }),
                // 打包的时候将被打包文件中的指定字符串替换为指定的值
                // https://github.com/rollup/rollup-plugin-replace
                replace({
                    exclude: 'node_modules/**',
                    CURRENT: JSON.stringify(process.env.NODE_ENV)
                }),
                istanbul({
                    exclude: ['test/**/*', 'node_modules/**/*']
                }),
                globals(),
                builtins(),
                nodeResolve({
                    jsnext: true,
                    extensions: ['.js', '.json']
                }),
                commonjs(),
                babel(babelrc({ addModuleOptions: false }))
            ],
            external: external,
            format: 'iife',
            moduleName: 'enjoy-version',// UMD、IIFE模式中需要 moduleName
            sourceMap: 'inline'
        },

        // 所用的测试结果报告器（reporter）
        // possible values: 'dots', 'progress'
        // 当前可用的报告器: https://npmjs.org/browse/keyword/karma-reporter
        reporters: [
            'progress',
            'coverage' // 这里使用karma-coverage这个报告器来生成覆盖率报告
        ],

        // 配置karma-coverage这个报告器，可选
        // coverageReporter: {
        //     type : 'html',
        //     dir : 'coverage/',
        //     // 仪表化器配置
        //     instrumenterOptions: {
        //         istanbul: { noCompact: true } // 不压缩
        //     }
        // },

        // 服务器端口
        port: 9876,

        // 控制输出的报告和log是否带有颜色
        colors: true,

        // 日志级别
        // 可选的值: config.LOG_DISABLE || config.LOG_ERROR || config.LOG_WARN || config.LOG_INFO || config.LOG_DEBUG
        logLevel: config.LOG_INFO,

        // 是否允许监听文件变化并在文件变化时自动执行测试
        autoWatch: true,

        // 要测试的目标环境
        // 可用的浏览器启动器（launcher）: https://npmjs.org/browse/keyword/karma-launcher
        browsers: ['Chrome'], // 'PhantomJS', 'Firefox', 'Chrome', 'ChromeCanary', 'Safari', 'Opera', 'IE'等

        // 持续集成模式
        // 如果设置singleRun: true，则测试完成后将关闭浏览器
        singleRun: false,

        // 并发级别
        // 需要同时打开多少个浏览器，infinity－无穷大
        concurrency: Infinity
    })
}


test/istanbul.reporter.js
const instanbul = require('istanbul');
const MochaSpecReporter = require('mocha/lib/reporters/spec');

module.exports = function (runner) {
    const collector = new instanbul.Collector();
    const reporter = new instanbul.Reporter();
    reporter.addAll(['lcov', 'json']);
    /* eslint-disable no-new */
    new MochaSpecReporter(runner);
    /* eslint-enable no-new */

    runner.on('end', () => {
        /* eslint-disable no-underscore-dangle */
        collector.add(global.__coverage__);
        /* eslint-enable no-underscore-dangle */

        reporter.write(collector, true, () => {
            process.stdout.write('report generated');
        });
    });
};


src/index.js
/**
 * 获取浏览设备相关信息的工具方法
 * @exports lib/enjoy-version
 * @author mobilesite
 * @email mobilesite@163.com
 * @since 2017-11-25
 */

import pkg from '../package.json';

export class Version {
    static compare(version1, version2) {
        const v1 = version1.toString().split('.');
        const v2 = version2.toString().split('.');
        const maxLen = Math.max(v1.length, v2.length);

        for (let i = 0; i < maxLen; i++) {
            let n1 = parseInt(v1[i], 10);
            let n2 = parseInt(v2[i], 10);

            isNaN(n1) && (n1 = 0);
            isNaN(n2) && (n2 = 0);

            if (n1 < n2) {
                return -1;
            } else if (n1 > n2) {
                return 1;
            }
        }
        return 0;
    }

    constructor(version) {
        if (version !== undefined && !isNaN(parseInt(version, 10))) {
            this.version = `${version}`; // 转成字符串格式
        } else {
            this.version = '';
        }
    }

    gt(version) {
        return Version.compare(this.version, version) > 0;
    }

    gte(version) {
        return Version.compare(this.version, version) >= 0;
    }

    lt(version) {
        return Version.compare(this.version, version) < 0;
    }

    lte(version) {
        return Version.compare(this.version, version) <= 0;
    }

    eq(version) {
        return Version.compare(this.version, version) === 0;
    }
}

export const libVersion = pkg.version;

/**
 * @property {string} libVersion - 当前模块的版本号
 * @property {Class} Version - 版本类，其中包含了
 */
export default {
    libVersion,
    Version
};


rollup.config.js
/**
 *
 * @author mobilesite
 * @email mobilesite@163.com
 * @since  2017-11-25
 */
'use strict';

console.log('building start:');
let rollup = require('rollup');
let replace = require('rollup-plugin-replace');
// Rollup的模块引用只支持 ES6 Module，其他的需要采用 npm 和 commonjs 的插件去解决
let json = require('rollup-plugin-json');// allows Rollup to import data = requirea JSON file
let nodeResolve = require('rollup-plugin-node-resolve');// teaches Rollup how to find external modules
let commonjs = require('rollup-plugin-commonjs');// the majority of packages on npm are exposed as CommonJS modules. We need to convert CommonJS to ES2015 before Rollup can process them.
let eslint = require('rollup-plugin-eslint');
let globals = require('rollup-plugin-node-globals');
let builtins = require('rollup-plugin-node-builtins');
let istanbul = require('rollup-plugin-istanbul');
let babel = require('rollup-plugin-babel');
let uglify = require('rollup-plugin-uglify');
let minify  = require('uglify-js').minify;
let babelrc = require('babelrc-rollup').default;
// let babelConfig  = require('./.babelrc.js');
let pkg = require('./package.json');
let external = Object.keys(pkg.dependencies);
let shell = require('shelljs');

const formatArr = [
    'es',
    'amd',
    'cjs',
    'iife',
    'umd'
];

shell.rm('-rf', 'dist');

// formatArr.map((item) => {
//     console.log(`building ${item} ...`);
//     rollup.rollup({
//         input: 'src/index.js',
//         plugins: [
//             json(),
//             eslint({ // https://www.npmjs.com/package/rollup-plugin-eslint
//                 throwError: true,
//                 exclude: ['package.json', 'node_modules/**']
//             }),
//             // Replace strings in files while bundling them.
//             // https://github.com/rollup/rollup-plugin-replace
//             replace({
//                 exclude: 'node_modules/**',
//                 CURRENT: JSON.stringify(process.env.NODE_ENV)
//             }),
//             istanbul({
//                 exclude: ['test/**/*', 'node_modules/**/*']
//             }),
//             nodeResolve({
//                 jsnext: true,
//                 extensions: ['.js', '.json']
//             }),
//             commonjs(),
//             babel(babelrc({ addModuleOptions: false })),// https://github.com/eventualbuddha/babelrc-rollup/issues/6
//             // https://github.com/TrySound/rollup-plugin-uglify
//             (process.env.NODE_ENV === 'production' && uglify(
//                 {
//                     output: {
//                         comments: function(node, comment) {
//                             let text = comment.value;
//                             let type = comment.type;
//                             if (type == "comment2") {
//                                 // multiline comment
//                                 return /@preserve|@license|@cc_on/i.test(text);
//                             }
//                         }
//                     }
//                 },
//                 minify
//             ))// 注意这里的minify，相应地需要在package.json中添加"uglify-js": "git+https://github.com/mishoo/UglifyJS2.git#harmony"，再重新install一下，解决uglify默认无法处理ES2015语法的问题
//         ],
//         external: external
//     }).then((bundle) => {
//         bundle.write({
//             format: item,
//             name: 'enjoy-version',// UMD、IIFE模式中需要 name
//             sourcemap: true,
//             file: item === 'es' ? `dist/enjoy-version.${item}.mjs` : `dist/enjoy-version.${item}.js`
//         });
//     }).catch(function(err){
//         // 注意下rollup的这种用catch方法捕获错误的方式，对于我们获知打包中的错误很有用。
//         // 而且，如果是自己写用 try catch 语句的方式，并不能捕获这些错误
//         console.log(err);
//     });
// });

module.exports = {
    input: 'src/index.js',
    plugins: [
        json(),
        eslint({ // https://www.npmjs.com/package/rollup-plugin-eslint
            throwError: true,
            exclude: ['package.json', 'node_modules/**']
        }),
        // Replace strings in files while bundling them.
        // https://github.com/rollup/rollup-plugin-replace
        replace({
            exclude: 'node_modules/**',
            CURRENT: JSON.stringify(process.env.NODE_ENV)
        }),
        istanbul({
            exclude: ['test/**/*', 'node_modules/**/*']
        }),
        globals(),
        builtins(),
        nodeResolve({
            jsnext: true,
            extensions: ['.js', '.json']
        }),
        commonjs(),
        babel(babelrc({ addModuleOptions: false })),// https://github.com/eventualbuddha/babelrc-rollup/issues/6
        // https://github.com/TrySound/rollup-plugin-uglify
        (process.env.NODE_ENV === 'production' && uglify(
            {
                output: {
                    comments: function(node, comment) {
                        let text = comment.value;
                        let type = comment.type;
                        if (type == 'comment2') {
                            // multiline comment
                            return /@preserve|@license|@cc_on/i.test(text);
                        }
                    }
                }
            },
            minify
        ))// 注意这里的minify，如果用的是uglify2，相应地需要在package.json中添加"uglify-js": "git+https://github.com/mishoo/UglifyJS2.git#harmony"，再重新install一下，解决uglify默认无法处理ES2015语法的问题
    ],
    external: external,
    output: [
        {
            format: 'es',
            name: 'enjoyVersion',// UMD、IIFE模式中需要 name
            sourcemap: true,
            file: 'dist/enjoyVersion.es.mjs'
        },
        {
            format: 'amd',
            name: 'enjoyVersion',// UMD、IIFE模式中需要 name
            sourcemap: true,
            file: 'dist/enjoyVersion.amd.mjs'
        },
        {
            format: 'cjs',
            name: 'enjoyVersion',// UMD、IIFE模式中需要 name
            sourcemap: true,
            file: 'dist/enjoyVersion.cjs.mjs'
        },
        {
            format: 'iife',
            name: 'enjoyVersion',// UMD、IIFE模式中需要 name
            sourcemap: true,
            file: 'dist/enjoyVersion.iife.mjs'
        },
        {
            format: 'umd',
            name: 'enjoyVersion',// UMD、IIFE模式中需要 name
            sourcemap: true,
            file: 'dist/enjoyVersion.umd.mjs'
        }
    ]
}
